src/Feldspar/Onnx/Operators.hs

```diff
@@ -442,3 +482,27 @@ vZipWith f vec1 vec2
 -- | Construct a Pull vector with an undefined index function and a given extent
 fromExt :: Shape sh -> Pull sh a
 fromExt = Pull (P.error "Extent of result depends on index function of argument.")
+
+{- The rationale for the definitions below is mainly max and min pooling in combination
+   with padding. ONNX specifies that the padding should not contribute to the max or min
+   value within a pooling window near the edge.
+-}
+
+class OnnxBounded a where
+  onnxMinBound :: a
+  onnxMaxBound :: a
+
+-- | OnnxBounded instances for Float
+instance OnnxBounded Float where
+  onnxMinBound = -(1/0)
+  onnxMaxBound =  (1/0)
+
+-- | OnnxBounded instance for Double
+instance OnnxBounded Double where
+  onnxMinBound = -(1/0)
+  onnxMaxBound =  (1/0)
+
+-- | OnnxBounded instance for Data
+instance (Hashable a, Type a, OnnxBounded a) => OnnxBounded (Data a) where
```

Comment:
Shouldn't there be a `Syntax` rather than `Hashable` and `Type` in this context?

Reply:
No, the type parameter of Data should be in the Type class (and in this case the Hashable class as well). If we had had just 'a' and not 'Data a' we would have had the context 'Syntax a'.

Reply:
To me there was a surprising amount of calls to `value` in this patch and historically that has led to bad code being generated. Are all the calls required? Is the code quality no longer an issue with the new CSE?

Unrelated to that, we're going to need instances for the integer and fixed point types in the future, which I'm guessing will be done with `instance Bounded a => OnnxBounded a`, does this design work with that?

Reply:
I have seen no particular problems with using 'value'. It simply generates a literal in the AST which is hashed just like any other. And since the variable name is the hash value, the variable the literal gets bound to becomes like a one place memo function so the hash value does not need to be recomputed. So I guess that that was a problem with the old CSE.

I think that such an instance would work. It's overlapping, but there are relatively good mechanisms to handle that. The type error messages will inevitably suffer somewhat, though.